src/cdk/schematics/update-tool/component-resource-collector.ts

```diff
@@ -0,0 +1,186 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {existsSync, readFileSync} from 'fs';
+import {dirname, resolve} from 'path';
+import * as ts from 'typescript';
+import {getAngularDecorators} from './utils/decorators';
+import {unwrapExpression} from './utils/functions';
+import {
+  computeLineStartsMap,
+  getLineAndCharacterFromPosition,
+  LineAndCharacter
+} from './utils/line-mappings';
+import {getPropertyNameText} from './utils/property-name';
+
+export interface ResolvedResource {
+  /** Class declaration that contains this resource. */
+  container: ts.ClassDeclaration|null;
+  /** File content of the given template. */
+  content: string;
+  /** Start offset of the resource content (e.g. in the inline source file) */
+  start: number;
+  /** Whether the given resource is inline or not. */
+  inline: boolean;
+  /** Path to the file that contains this resource. */
+  filePath: string;
+  /**
+   * Gets the character and line of a given position index in the resource.
+   * If the resource is declared inline within a TypeScript source file, the line and
+   * character are based on the full source file content.
+   */
+  getCharacterAndLineOfPosition: (pos: number) => LineAndCharacter;
+}
+
+/**
+ * Collector that can be used to find Angular templates and stylesheets referenced within
+ * given TypeScript source files (inline or external referenced files)
+ */
+export class ComponentResourceCollector {
```

Comment:
A lot of this code seems like stuff that would fit neatly into the devkit. What are your thoughts on eventually reintegrating it?

Reply:
Totally agree. We need to think about this more and should maybe have a discussion about this. Same applies for various utilities in `@angular/core/schematics`. 

I'd consider exposing stuff like the `update-tool`. At least in a common place so we can use it for `@angular/core` migrations as well (and CLI migrations etc.)